Light Freshness Edit: ASP.NET - Minimal APIs (#37060)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/minimal-apis/security.md

---
title: Authentication and authorization in Minimal APIs
author: wadepickett
description: Learn how to configure authentication and authorization in Minimal API apps, explore concepts, define policies, and run development tests.
ms.author: wpickett
content_well_notification: AI-contribution
monikerRange: '>= aspnetcore-7.0'
ms.date: 04/28/2026
uid: fundamentals/minimal-apis/security
ai-usage: ai-assisted

# customer intent: As an ASP.NET developer, I want to use authentication and authorization in Minimal API apps, so I can manage user access to my apps.
---

# Authentication and authorization in Minimal APIs

[!INCLUDE[](~/includes/not-latest-version.md)]

Minimal APIs support all authentication and authorization options available in ASP.NET Core, and provide extra functionality to improve the experience for managing authentication.

This article describes the support for authentication and authorization in Minimal API applications, and how to configure and test the functionality.

> ## Compare authentication and authorization 

Reviewer: ```suggestion
## Authentication and authorization concepts
```
Minor item: These are not being compared really, maybe this would be better.

### Authorization service

In the authorization layer, there are two strategies for determining user access to resources:

- **Role-based strategies** determine a user's access based on their assigned role, such as `Administrator` or `User`. For more information on role-based authorization, see [role-based authorization documentation](../../security/authorization/roles.md).

- **Claim-based strategies** determine a user's access based on claims issued by a central authority. For more information on claim-based authorization, see [claim-based authorization documentation](../../security/authorization/claims.md).

> In ASP.NET Core, both strategies are captured into an authorization requirement. The authorization service exercises authorization handlers to determine whether or not a particular user meets the authorization requirements to a resource. 

Reviewer: ```suggestion
In ASP.NET Core, both strategies are captured into an authorization requirement. The authorization service uses authorization handlers to determine whether or not a particular user meets the authorization requirements for a resource. 
```
Minor item:  I think "uses" would be better here.  The authorization service doesn't "exercise" handlers in this context, it uses them.

## Enable authentication in minimal apps

> To enable authentication, call the [AddAuthentication](/dotnet/api/microsoft.extensions.dependencyinjection.authenticationservicecollectionextensions.addauthentication) to register the required authentication services on the app's service provider.

Reviewer: ```suggestion
To enable authentication, call [AddAuthentication](/dotnet/api/microsoft.extensions.dependencyinjection.authenticationservicecollectionextensions.addauthentication) to register the required authentication services on the app's service provider.
```

Minor item: This can be either call AddAuthentication" (no article) or "call the AddAuthentication method."

:::code language="csharp" source="~/fundamentals/minimal-apis/security/7.0-samples/MinApiAuth/MinApiAuth/Program.cs" id="snippet_jwt2":::

In some cases, such as controlling middleware order, it's necessary to explicitly register authentication and authorization. In the following sample, the authentication middleware runs _after_ the CORS middleware runs. For more information on middleware and this automatic behavior, see [Middleware in Minimal API apps](./middleware.md).

:::code language="csharp" source="~/fundamentals/minimal-apis/security/7.0-samples/MinApiAuth/MinApiAuth/Program.cs" id="snippet_after" highlight="9-11":::

### Configure authentication strategy

Authentication strategies typically support various configurations that are loaded via options. Minimal apps support loading options from configuration for the following authentication strategies:

* [JWT bearer-based authentication](https://jwt.io/introduction#what-is-json-web-token)

> * [OpenID Connect, connection-based authentication](https://openid.net/developers/how-connect-works/)

Reviewer: ```suggestion
* [OpenID Connect-based authentication](https://openid.net/developers/how-connect-works/)
```
Thanks for catching the orignial typo.  Minor-item: I think this version would be a little more clear.  See what you think.

The scheme name is used to uniquely identify an authentication strategy. The name is also used as the lookup key when resolving authentication options from config, as shown in the following example:

:::code language="csharp" source="~/fundamentals/minimal-apis/security/7.0-samples/MinApiAuth/MinApiAuth/Program.cs" id="snippet_local" highlight="5":::

## Configure authorization policies in minimal apps

Authentication identifies and validates the identity of users against an API. Authorization validates and verifies access to resources in an API. The `IAuthorizationService` service registered by the `AddAuthorization` extension method facilitates the authorization. In the following scenario, a `/hello` resource is added that requires a user to present an `admin` role claim with a `greetings_api` scope claim.

Configuration of the authorization requirements on a resource is a two-step process:

1. Define the authorization requirements in a policy globally.
1. Apply individual policies to resources.

> In the following code, the <xref:Microsoft.Extensions.DependencyInjection.PolicyServiceCollectionExtensions.AddAuthorizationBuilder%2A> is invoked which:

Reviewer: ```suggestion
In the following code, <xref:Microsoft.Extensions.DependencyInjection.PolicyServiceCollectionExtensions.AddAuthorizationBuilder%2A> is invoked, which:
```
Very minor item: Seemed a tiny bit awkward, see if that seems better.
